traffic_ops/traffic_ops_golang/test/error_context.go

```diff
@@ -0,0 +1,267 @@
+/*
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package test
+
+import "fmt"
+
+type errorCode struct {
+	error
+	cause error
+	code  int
+}
+
+type ErrorCoder interface {
+	Error() string
+	ErrorCode() int
+	Cause() error
+	Prepend(string, ...interface{}) ErrorCoder
+}
+
+func NewError(code int, fmtStr string, fmtArgs ...interface{}) ErrorCoder {
+	err := fmt.Errorf(fmtStr, fmtArgs...)
+	return &errorCode{err, err, code}
+}
+
+func (e errorCode) ErrorCode() int {
+	return e.code
+}
+
+// not a pointer receiver, does not modify itself
+// in the new error the original cause is maintained
+func (e errorCode) Prepend(fmtStr string, fmtArgs ...interface{}) ErrorCoder {
+	err := fmt.Errorf(fmtStr, fmtArgs...)
+	e.error = fmt.Errorf("%v %v", err, e.error)
+	return e
+}
+
+func (e errorCode) Cause() error {
+	return e.cause
+}
+
+func AddErrorCode(code int, err error) ErrorCoder {
+	return NewError(code, "%v", err)
+}
+
+// ErrorContext regulates which error codes can be made and keeps a count
+// of all the errors created through the context
+//
+// ErrorContext.NewError can be used like test.NewError
+//
+// Implementation details:
+// contains a list of all valid error codes
+//		- allows user to make sure they are creating the correct errors
+//		- actually a map
+//			lookup can be done without linear search
+//			we can use the map to keep count of which errors are made
+// contains mapping from error code to name (either for testing metainfo or used in case no args are given)
+//		- not required for all error codes, or for any
+//
+// Recommended setup:
+//
+//	package some_regex_checker
+//
+//	const (
+//		COMMON_BASE            = 10 + iota
+//		NOT_ENOUGH_ASSIGNMENTS
+//		BAD_ASSIGNMENT_MATCH
+//		...
+//	)
+//
+//	// scoped to the package name
+//	var ErrorContext *test.ErrorContext
+//
+//	func init() {
+//		errorCodes := []uint{
+//			NOT_ENOUGH_ASSIGNMENTS,
+//			BAD_ASSIGNMENT_MATCH,
+//		}
+//		ErrorContext = test.NewErrorContext("cache config", errorCodes)
+//		err := ErrorContext.AddMapping(NOT_ENOUGH_ASSIGNMENTS, "not enough assignments in rule")
+//		// check err
+//		ErrorContext.TurnPanicOn()
+//		// in non-init function create errors with the context
```

Comment:
>create errors with the context

How? Would you mind adding another example of how to do that? Should that example be in a GoDoc comment for `ErrorCoder`, and here `// in non-init function create errors with the context - see ErrorCoder` maybe?

Reply:
I expanded the example so that a comment isn't used.